src/rgw/driver/posix/rgw_sal_posix.h

```diff
@@ -0,0 +1,718 @@
+  virtual int check_bucket_shards(const DoutPrefixProvider* dpp, optional_yield y) override;
+  virtual int chown(const DoutPrefixProvider* dpp, User& new_user, optional_yield y) override;
+  virtual int put_info(const DoutPrefixProvider* dpp, bool exclusive,
+		       ceph::real_time mtime, optional_yield y) override;
+  virtual int check_empty(const DoutPrefixProvider* dpp, optional_yield y) override;
+  virtual int check_quota(const DoutPrefixProvider *dpp, RGWQuota& quota, uint64_t obj_size, optional_yield y, bool check_size_only = false) override;
+  virtual int try_refresh_info(const DoutPrefixProvider* dpp, ceph::real_time* pmtime, optional_yield y) override;
+  virtual int read_usage(const DoutPrefixProvider *dpp, uint64_t start_epoch,
+			 uint64_t end_epoch, uint32_t max_entries, bool* is_truncated,
+			 RGWUsageIter& usage_iter, std::map<rgw_user_bucket, rgw_usage_log_entry>& usage) override;
+  virtual int trim_usage(const DoutPrefixProvider *dpp, uint64_t start_epoch, uint64_t end_epoch, optional_yield y) override;
+  virtual int remove_objs_from_index(const DoutPrefixProvider *dpp, std::list<rgw_obj_index_key>& objs_to_unlink) override;
+  virtual int check_index(const DoutPrefixProvider *dpp, std::map<RGWObjCategory, RGWStorageStats>& existing_stats, std::map<RGWObjCategory, RGWStorageStats>& calculated_stats) override;
+  virtual int rebuild_index(const DoutPrefixProvider *dpp) override;
+  virtual int set_tag_timeout(const DoutPrefixProvider *dpp, uint64_t timeout) override;
+  virtual int purge_instance(const DoutPrefixProvider* dpp, optional_yield y) override;
+
+  virtual std::unique_ptr<Bucket> clone() override {
+    return std::make_unique<POSIXBucket>(*this);
+  }
+
+  virtual std::unique_ptr<MultipartUpload> get_multipart_upload(
+				const std::string& oid,
+				std::optional<std::string> upload_id=std::nullopt,
+				ACLOwner owner={}, ceph::real_time mtime=real_clock::now()) override;
+  virtual int list_multiparts(const DoutPrefixProvider *dpp,
+			      const std::string& prefix,
+			      std::string& marker,
+			      const std::string& delim,
+			      const int& max_uploads,
+			      std::vector<std::unique_ptr<MultipartUpload>>& uploads,
+			      std::map<std::string, bool> *common_prefixes,
+			      bool *is_truncated, optional_yield y) override;
+  virtual int abort_multiparts(const DoutPrefixProvider* dpp,
+			       CephContext* cct, optional_yield y) override;
+
+  /* Internal APIs */
+  int create(const DoutPrefixProvider *dpp, optional_yield y, bool* existed);
+  void set_stat(struct statx _stx) { stx = _stx; stat_done = true; }
+  int get_dir_fd(const DoutPrefixProvider *dpp) { open(dpp); return dir_fd; }
+  /* TODO dang Escape the bucket name for file use */
+  std::string get_fname();
+  int get_shadow_bucket(const DoutPrefixProvider* dpp, optional_yield y,
+			const std::string& ns, const std::string& tenant,
+			const std::string& name, bool create,
+			std::unique_ptr<POSIXBucket>* shadow);
```

Comment:
I did it this way to avoid code duplication.

Reply:
understood, but you can share code without sharing the interface. i'm just picturing a `class POSIXBucket` and `class MultipartDirectory` that both call helper functions that operate on file descriptors and paths

with this separation, it would be much easier to make changes or bugfixes to one without introducing weird regressions in the other